[AIRFLOW-3205] Support multipart uploads to GCS

chronitis opened the pull request:
Cloud Storage supports resumable/multipart uploads for large files,
which can be used to avoid limitations on the size of a single HTTP
request, or by adding a retry behaviour, increase the reliability of
large transfers.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title:
  - https://issues.apache.org/jira/browse/AIRFLOW-3205

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

The current `gcs_hook` only supports single-part uploads, which limits the ultimate size of any single file upload to that which the SSL library supports in a single request, and may also lead to issues where the network is not wholly reliable and very long-living connections may be more likely to fail. This adds support for multipart uploads (uploading a single file, with the actual transport split into multiple HTTP requests), and retries (using the retry functionality from the `google-api-python-client` library).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Tests for upload behavior (both multipart and single part) have been added to `tests.contrib.hooks.test_gcs_hook`.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

The added arguments are documented in the hook `upload` method. I haven't explicitly added all these arguments to classes which use the hook.

### Code Quality

- [x] Passes `flake8`

Commit:
[AIRFLOW-3205] Support multipart uploads to GCS

Cloud Storage supports resumable/multipart uploads for large files,
which can be used to avoid limitations on the size of a single HTTP
request, or by adding a retry behaviour, increase the reliability of
large transfers.

XD-DENG commented on airflow/contrib/hooks/gcs_hook.py:
```diff
@@ -191,6 +193,15 @@ def upload(self, bucket, object, filename,
+        :param multipart: If true, split the upload into multiple HTTP requests
+        :type multipart: bool
+        :param chunksize: Size in bytes of each upload chunk when performing a
+                          multipart upload; must be a multiple of 256KiB (262144 bytes)
```
May be good to mention your default value (256Mb) here in the docstring?

Fokko commented on airflow/contrib/hooks/gcs_hook.py:
```diff
@@ -177,7 +177,9 @@ def download(self, bucket, object, filename=None):
     def upload(self, bucket, object, filename,
-               mime_type='application/octet-stream', gzip=False):
+               mime_type='application/octet-stream', gzip=False,
+               multipart=False, chunksize=256 * 1024 * 1024,
```
API Design question. It would also be possible to set chunksize to None by default, it set then it will switch to multipart. This will reduce the number of options (because `multipart` is set implicitly). What do you think?

chronitis replied:
Avoiding two options seems beneficial, but the onus is then on every call site to specify not just that the upload should be multipart, but explicitly select a size.

Would this work

```
upload(..., multipart=False)  # default, single-part upload
upload(..., multipart=True)  # multipart with default chunksize
upload(..., multipart=1024 * 1024)  # multipart with chosen chunksize
```

with `false` being the default value in the signature?

Commit:
[AIRFLOW-3205] Use only the multipart keyword

This removes the chunksize keyword, using instead multipart=True for a
default chunk size or multipart=int to override the default.

Fokko reviewed (approved):
Yes, thank you @chronitis. Merging to master
